framework/include/utils/MooseTypes.h

```diff
@@ -270,6 +270,8 @@ struct VariableValueType
   typedef
       typename OutputTools<typename Moose::ValueType<T, compute_stage>::type>::VariableValue type;
 };
+typedef typename VariableValueType<Real, ComputeStage::JACOBIAN>::type DualVariableValue;
```

Comment:
Have you tried out the `using` syntax for these typedefs? Just curious. I wonder if you could drop the typename part with the newer syntax.

Reply:
Actually the `typename` here is extraneous since `VariableValueType<Real, ComputeStage::JACOBIAN>::type` is not actually dependent on any template parameters.

`using` vs `typedef` keywords have the same requirements for when `typename` has to be used, but `using` is more versatile because it allows you to create aliases even for templates. `typedef` is limited to the case where all template arguments are already available.

Just some examples from me playing around :smile:

```c++
template <typename T> class B {
public:
  typedef T BType;
};

template <typename T> class A {
public:
  typedef B<T> AType;
  typedef typename B<T>::BType
      AType2; // Need typename because BType is dependent on `T`

  template <typename U>
  using SomeType =
      typename B<U>::BType; // Need typename because BType is dependent on `U`
};

typedef A<int>::AType MainType;
using MainType2 = A<int>::AType;

template <typename T>
using TemplateType = typename A<T>::AType; // Again, need typename because AType
                                           // is dependent on `T`

using AnotherType = B<int>;

int main() {}
```